### ISSUE TITLE
fluctuations in evaluation metrics during training 

### ISSUE THREAD

**Issue by kirk86** [cornellius-gp/gpytorch#882]
# 📚 Documentation/Examples

** Is there documentation missing? **
No

** Is documentation wrong? **
No

** Is there a feature that needs some example code? **
No

** Clarification on fluctuation metrics in deep kernel example **
Yes

Hi folks and thanks for making gpytorch available! I was wondering if anyone could shed some light on the following pattern that is exhibit from the deep kernel example tutorial. Basically I see a lot of fluctuation of accuracy going from 90% to 50% to 80%.....

![image](https://user-images.githubusercontent.com/2902390/65626321-3313d580-dfc5-11e9-9e23-e670a02bdb92.png)

1. Is it normal for deep GPs to have these fluctuations?

2. Any ideas what might lead to cause these fluctuations? (e.g. normal deep learning models usually don't fluctuate that much)?

3. I had a really hard time getting the deep kernel model to go anything above 90% accuracy on cifar10 (i.e. only with data augmentation). Any ideas on improvements? Maybe different likelihoods or kernels would be better?

Thanks!

**Comment by jacobrgardner** [comment]
@kirk86 In the first 150 epochs before the first learning rate drop, the test error of vanilla densenets also fluctuates quite wildly. See e.g. Figure 4 in the paper:
![image](https://user-images.githubusercontent.com/4016393/65635677-9e45b380-df95-11e9-93d8-b1c130efa606.png).

In general, I don't think it's too surprising to see fluctuations mostly between 15% and 10% error with a few larger spikes. The hyperparameters in the example notebook are directly taken from the DenseNet paper, and can definitely be optimized more for DKL specifically, but the accuracy should essentially match after completing training.

**Comment by kirk86** [comment]
@jacobrgardner Hey Jake, thanks for the pointers and clarifications. I'll try swapping densenet for a resnet and see what happens. If you don't mind me asking another question. I've read numerous papers claiming that GPs are good at handling uncertainty but empirical I haven't evidenced this. For instance comparing the deepkernel model against a vgg16-sgd it seems to be doing awfully worse than the plain sgd vgg16 and I couldn't find a justification for that, any ideas?

**Comment by gpleiss** [comment]
@kirk86 there's a lot of reasons that this could be happening. What exactly do you mean by worse? (Error? Log likelihood? Uncertainties?)

SV-DKL uses an approximate GP (since inference is intractable). The more extreme the approximations (e.g. fewer inducing points), the more likely it is to have some effect on the output. Additionally, adding a GP on top of the NN does change the optimization landscape. From my own experience, you do have to play around a bit with the GP/network/optimization hypers to get good optimization, depending on the network.

I would play around with the network architecture, as well as some of the GP hyperparameters. You should be seeing *at least* similar performance to the plain NN architecture.

**Comment by kirk86** [comment]
@gpleiss Hi Geoff, thanks for the reply.

> What exactly do you mean by worse? (Error? Log likelihood? Uncertainties?)

Allow me to elaborate a bit, in terms of accuracy the SV-DKL is 2%-3% lower than plain NN on test set but that isn't much of a difference if you ask me. On the other hand it seems that is probably the worst calibrated one in terms of ECE and reliability plots also in terms of uncertainties.

> Additionally, adding a GP on top of the NN does change the optimization landscape.

It does make sense since I was thinking that because it includes an NN inside, SV-DKL implicity inherits some evident problems in NN, but I was kind of expecting the GP on top to mitigate for some of those? I guess I was wrong...

**Comment by andrewgordonwilson** [comment]
It can help, but the details matter. You could start by pre-training the base NN, and then starting from that initialization in training the DKL model. It's unlikely that would do worse than the NN.

**Comment by kirk86** [comment]
@andrewgordonwilson That's something I haven't tied yet. It seems worth giving it a shot. Thank you for the advice I'll look into it!

**Comment by gpleiss** [comment]
Closing for now, but reopen if you have more questions :)
